### ISSUE TITLE
Add an `Object.groupBy` TypeScript goody

### ISSUE THREAD

**Issue by miorel** [code-chronicles-code/leetcode-curriculum#363]
An [`Object.groupBy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/groupBy) method is being added to JavaScript, but for now this is not available on LeetCode, since it requires Node 21, whereas LeetCode uses Node 20.

We'd like to get access to this function sooner, so let's create our own version!

This issue involves adding another directory within `workspaces/adventure-pack/goodies/typescript`, named `Object.groupBy`, with an `index.ts` for the implementation and an `index.test.ts` for a test. You can use the `Map.groupBy` added by #358 for inspiration, this should be very similar.

Note: after adding or updating a goody you will need to run `yarn package-goodies:test -u` to refresh the snapshot tests.

Please comment with any questions.

**Comment by elimanzo** [comment]
@miorel Quick question would this issue be the same as #108

**Comment by miorel** [comment]
Not exactly. This one is about replicating the built-in function `Object.groupBy`, #108 would be about adding a method to iterators.
